src/EditorFeatures/CSharp/CompleteStatement/CompleteStatementCommandHandler.cs

```diff
@@ -0,0 +1,397 @@
+
+        public void ExecuteCommand(TypeCharCommandArgs args, Action nextCommandHandler, CommandExecutionContext executionContext)
+        {
+            BeforeExecuteCommand(args, executionContext);
+            nextCommandHandler();
+        }
+
+        private void BeforeExecuteCommand(TypeCharCommandArgs args, CommandExecutionContext executionContext)
+        {
+            if (args.TypedChar != ';')
+            {
+                return;
+            }
+
+            var caret = args.TextView.GetCaretPoint(args.SubjectBuffer);
+            if (!caret.HasValue)
+            {
+                return;
+            }
+
+            var isCaretAtEndOfLine = ((SnapshotPoint)caret).Position == ((SnapshotPoint)caret).GetContainingLine().End;
+            var document = caret.Value.Snapshot.GetOpenDocumentInCurrentContextWithChanges();
+            if (document == null)
+      
+            {
+                return;
+            }
+
+            var root = document.GetSyntaxRootSynchronously(executionContext.OperationContext.UserCancellationToken);
+            var caretPosition = caret.Value.Position;
+
+            var token = root.FindToken(caretPosition);
+
+            var currentNode = token.Parent;
+
+            // If cursor is right before an opening delimiter, start with node outside of delimiters. 
+            // This covers cases like `obj.ToString$()`, where `token` references `(` but the caret isn't actually 
+            // inside the argument list.
+            if (token.IsKind(SyntaxKind.OpenBraceToken, SyntaxKind.OpenBracketToken, SyntaxKind.OpenParenToken)
+                && token.Span.Start >= caretPosition)
+            {
+                currentNode = currentNode.Parent;
+            }
+
+            if (currentNode == null)
+            {
+                return;
+            }
+
+            var syntaxFacts = document.GetLanguageService<ISyntaxFactsService>();
+            if (!LooksLikeNodeInArgumentListForStatementCompletion(currentNode, syntaxFacts, isCaretAtEndOfLine))
+            {
+                return;
+            }
+
+            // verify all delimeters exist until you reach statement syntax that requires a semicolon
+            while (!IsStatementOrFieldDeclaration(currentNode, syntaxFacts))
+            {
+                if (!ClosingDelimiterExistsIfNeeded(currentNode))
+                {
+                    // A required delimiter is missing; do not treat semicolon as statement completion
+                    return;
+                }
+
+                if (currentNode.Parent == null)
+                {
+                    return;
+                }
+
+                currentNode = currentNode.Parent;
+            }
+
+            // if the statement syntax itself requires a closing delimeter, verify it is there
+                if (StatementClosingDelimiterIsMissing(currentNode))
+                {
+                    // Example: missing final `)` in `do { } while (x$$`
```

Comment:
i don't really get teh examples.  They mention a case, but they don't mention *why* that case should behave this way.  i.e. i don't understand why we want to do things differently for that do/while case.

Reply:
This is a nuance of the syntax trees, but I'm not sure we need to expand on the comments more than we already do for the following reasons:

1. We have tests that verify this block is included. If someone thinks it might be unnecessary and removes it, they'll quickly find out it's required.
2. The comments already point out the relevant example. Most closing delimiters are checked a few lines up, for the case where it's part of an expression and not the statement itself. However, in some cases, a statement has delimiters as an immediate child, so they need to be checked as part of the statement. One example is a do/while statement, for which the parentheses following `while` are children of the same syntax node as the `;`.
3. Eventually we'd like to expand the whole thing to also handle cases where closing delimiters are missing. (i.e. we already point out where we check for delimiters, and have a future follow-up to revisit this specific part of the design.